### ISSUE TITLE
the plan for sablejs 2.0

### ISSUE THREAD

**Issue by ErosZy** [ErosZy/sablejs#19]
sablejs 1.x has been running well on [YoTest](https://www.fastyotest.com/), but because of the private opcode design, we don't open all code for community(preventing decompiling). After discussions, sablejs 2.0 will plan to open all code! To reach this goal, we will make a major upgrade to the current code of 1.x, include:

**1. remove the private opcode design and transform for the stack operation directly**
**2. dynamically import inner object for sandbox**

### 1. Transform Directly

Since sablejs 1.x is fully stack-based vm, we will have a big loop to execute the relevant opcode code. However, this approach causes a significant performance loss due to the failure of branch prediction: 
```javascript
for (;;) {
  switch (opcode) {
    case 0: {
      // ...
    }
    case 1: {
      // ...
    }
    case 2: {
      // ...
    }
  }
}
```
After using the latest d8 and perf for profiling, about 30% of the performance consumption is wasted. To solve this problem, and to make V8 better able to help us with JIT, we will perform the equivalent transform operation directly, consider the following example:
```javascript
function add() {
  return 1 + 2 + 3;
}
```
In 1.x, we will get linear opcode bytes, which will then be executed by vm. But in 2.0, we will compile directly to the following code:
```javascript
function __C_add(J) {
  __pushNumber(J, 1);
  __pushNumber(J, 2);
  __R_add(J);
  __pushNumber(J, 3);
  __R_add(J);
}
```
Here, stack manipulation methods such as `__pushNumber` are still provided by runtime. In this way, together with the relevant compilation optimizations in 1.x, there will be a relatively large performance improvment for frequent execution(benchmark can be followed by [DoppioJVM Web JIT implement](https://github.com/plasma-umass/doppio/pull/440)).

```javascript
(function(){
  for(var i = 0; i < 10000000; i++);
}());

// sablejs 2.0: 276.279ms --- baseline
// sablejs 1.0.6 878ms --- slower: 218.11%
// quickjs-wasm: 228ms --- faster: 17.39%
```

**At the same time, this brings the benefit of not having to rely on opcode, so we can directly open all of sablejs's code.** 

### 2. Dynamically Import Inner Object

In most of the usage of the current feedback, it is basically using sablejs for JSVMP. But sablejs 1.x is mainly designed with sandbox as the core, after 2.0 we will mainly aim at JSVMP while taking into account the functionality of sandbox,(the Inner Object will be import on demand according to your needs).

**Also, when you are using JSVMP only, sablejs 2.0 will get a very big performance improvment thanks to V8's object optimization!**

I considered for releasing sablejs 2.0 in mid-2022, please look forward to it! 😁

**Comment by fabiospampinato** [comment]
Will v2 include the ability to evaluate a raw javascript string without compiling it at build times first? That seems a big limitation at the moment if the goal of the project is to be able to be used as a secure vm for plugins.

**Comment by ErosZy** [comment]
@fabiospampinato If the compiler is included, then there is a risk of being decompiled, and there should be no plans to put the compiler and interpreter together at this time.

**Comment by npc1054657282** [comment]
Hello, will sablejs-2.0 support some new features in ES2016+?
I'm a ES2016+ user. I'd like to try the product, but some features like 'bigint' cannot be translated to ES2015 by babel.

**Comment by ErosZy** [comment]
@npc1054657282 maybe not, i have little time to maintance this project, but you can add es2016+ features by yourself.

**Comment by Foolyou** [comment]
Great work! I have two questions:
Is the code transformed by the direct transform method eventually executed by an eval(or new Function call)? Does this mean we cannot use sable2.x in an eval-forbidding environment?

**Comment by ErosZy** [comment]
> Great work! I have two questions: Is the code transformed by the direct transform method eventually executed by an eval(or new Function call)? Does this mean we cannot use sable2.x in an eval-forbidding environment?

No, it still run with stack ops, don't need eval or new Funcation. But maybe the schedule will be late, i'm really really busy and have no passion for 2.0 implementation. At the same time, i learn some optimization from v8, i think it will faster than i think before. But it needs some time.

**Comment by AngusFu** [comment]
Any progress?

**Comment by 250king** [comment]
I have same question about the progress(

**Comment by HackerYunen** [comment]
有生之年系列+1 😭

**Comment by ErosZy** [comment]
@250king @HackerYunen @AngusFu @Foolyou @fabiospampinato @npc1054657282 

Hi guys, **sablejs 2.0.0-beta is officially released! 🎉🎉🎉**

> But please note that this is still a very early release, so there may be some rough edges in the developer experience, even though we have already passed the vast majority of our test cases.

With major optimizations to AOT compilation and the host object model, sablejs has made significant improvements in both usability and performance. At the same time, in the AI era, we have also redefined the positioning of sablejs. It now focuses on:

- **user plugins**
- **rule/formula engines**
- **AI-generated JavaScript**
- **explicit capabilities + Worker isolation**
- **AOT compilation to direct JavaScript instead of runtime interpretation**

Performance, sandbox semantics, Test262 coverage, fuzzing, and the new Worker API are documented in the updated [README](https://github.com/ErosZy/sablejs). If you were interested in the original sablejs, I'd really appreciate it if you could try the new version.

```js
const fs = require("node:fs");
const { compile } = require("sablejs");
const generated = compile("({ total: input.price * 1.2 });");
fs.writeFileSync("program.cjs", generated.code);
```

```js
const program = require("./program.cjs");
const instance = program.createInstance({
  globals: { input: { price: 100 } },
});

try {
   // { total: 120 } — synchronous, returns the value
  console.log(instance.run());
} finally {
  instance.dispose();
}
```

Three-run medians on the Linux x64 reference machine (methodology, exclusions, and variance: [Performance](docs/performance.md)):

| Backend | V8 Benchmark Suite 7 score | vs sandbox |
| --- | ---: | ---: |
| sablejs O2 sandbox | 2,202 | — |
| sablejs O2 trusted | 2,783 | 1.26x |
| QuickJS-WASM 0.32.0 | 1,181 | 0.54x |

| Suite (direction) | sablejs trusted | sablejs sandbox | QuickJS-WASM |
| --- | ---: | ---: | ---: |
| Octane 2.0 geometric score (higher is better) | 2,674 | 2,088 | 1,472 |
| SunSpider 1.0 total, 23 tests (ms, lower is better) | 396.1 | 515.5 | 597.7 |
| Kraken 1.1 total, 14 tests (ms, lower is better) | 5,370.9 | 15,935.3 | 27,081.9 |

Feedback on real plugin/sandbox use cases, API design, performance, or security would be especially valuable.
